controllers/cruisecontroltask_controller.go

```diff
@@ -70,121 +82,223 @@ func (r *CruiseControlTaskReconciler) Reconcile(ctx context.Context, request ctr
 			if task == nil {
 				continue
 			}
-			task.FromResult(result)
+			task.SetCruiseControlOperationRef(cruiseControlOpRef)
+			task.SetStateScheduled()
 		}
-
-	case tasksAndStates.NumActiveTasksByOp(OperationRemoveBroker) > 0:
+	case tasksAndStates.NumActiveTasksByOp(banzaiv1alpha1.OperationRemoveBroker) > 0:
 		var removeTask *CruiseControlTask
-		for _, task := range tasksAndStates.GetActiveTasksByOp(OperationRemoveBroker) {
+		for _, task := range tasksAndStates.GetActiveTasksByOp(banzaiv1alpha1.OperationRemoveBroker) {
 			removeTask = task
 			break
 		}
 
 		details := []interface{}{"operation", "remove broker", "brokers", removeTask.BrokerID}
 
-		result, err := scaler.RemoveBrokers(removeTask.BrokerID)
+		cruiseControlOpRef, err := r.removeBroker(ctx, instance, operationTTLSecondsAfterFinished, removeTask.BrokerID)
 		if err != nil {
-			log.Error(err, "removing broker(s) from Kafka cluster via Cruise Control failed", details...)
+			log.Error(err, "creating CruiseControlOperation for downscale has failed", details...)
 		}
-		removeTask.FromResult(result)
+		removeTask.SetCruiseControlOperationRef(cruiseControlOpRef)
+		removeTask.SetStateScheduled()
 
-	case tasksAndStates.NumActiveTasksByOp(OperationRebalanceDisks) > 0:
-		logDirsByBroker, err := scaler.LogDirsByBroker()
+	case tasksAndStates.NumActiveTasksByOp(banzaiv1alpha1.OperationRebalance) > 0:
+		logDirsByBroker, err := r.Scaler.LogDirsByBroker()
 		if err != nil {
 			log.Error(err, "failed to get list of volumes per broker from Cruise Control")
 			return requeueAfter(DefaultRequeueAfterTimeInSec)
 		}
-
 		brokerIDs := make([]string, 0)
-		for _, task := range tasksAndStates.GetActiveTasksByOp(OperationRebalanceDisks) {
-			if task == nil || task.IsDone() {
-				continue
-			}
+		unavailableBrokerIDs := make([]string, 0)
+		for _, task := range tasksAndStates.GetActiveTasksByOp(banzaiv1alpha1.OperationRebalance) {
+			brokerIDs = append(brokerIDs, task.BrokerID)
+			found := false
 			if onlineDirs, ok := logDirsByBroker[task.BrokerID][scale.LogDirStateOnline]; ok {
-				found := true
 				for _, dir := range onlineDirs {
-					if !strings.HasPrefix(strings.TrimSpace(dir), strings.TrimSpace(task.Volume)) {
-						found = false
+					if strings.HasPrefix(strings.TrimSpace(dir), strings.TrimSpace(task.Volume)) {
+						found = true
 					}
 				}
-				if found {
-					brokerIDs = append(brokerIDs, task.BrokerID)
+				if !found {
+					unavailableBrokerIDs = append(unavailableBrokerIDs, task.BrokerID)
 				}
-			} else {
-				task.Err = "log dir is not available in Cruise Control"
 			}
 		}
+		if len(unavailableBrokerIDs) > 0 {
+			log.Info("requeue as there are offline broker log dirs for rebalance", "brokerIDs", unavailableBrokerIDs)
+			// This requeue is not necessary because the cruiseContrlOperation controller retries the errored task
+			return requeueAfter(DefaultRequeueAfterTimeInSec)
```

Comment:
In this case there will be completedWithError in the kafkaCluster status until the all of the brokers become available.
There will be requeue in the CruiseControlOperation reconciler or here.
I think better to do here because there will be no completedWithError status

Reply:
hmmmm if that is the case, I think we should update the comment to reflect this so people that look at this piece of code will have some context

Reply:
@bartam1 please update the comment in the source code to reflect your response in this thread